### ISSUE TITLE
Material distance fade is based on depth not distance

### ISSUE THREAD

**Issue by elvisish** [godotengine/godot#53853]
### Godot version

3.3.3

### System information

Windows 7

### Issue description

The spatial material distance fade is dependent on the facing angle of the camera, not the distance of the object. I believe this is because the distance fade is going on the depth rather than the distance:
![image](https://user-images.githubusercontent.com/16231628/137530827-46e8b931-5cd3-4193-a0b8-86362800a7c8.png)
This can be an issue in particular if distance fade is used to obscure objects from the player unless very close to them, as the distance of visibility changes drastically depending on the angle it's viewed from.

### Steps to reproduce

Set the distance fade on a spatial material, rotate the camera and observe it become less faded the closer to the edge of the screen and more faded the closer to the center of the screen.

**Comment by elvisish** [comment]
Thanks for adding as a milestone, and great to see there is a solution! Can materials be extended in their current form so I can get them to work like this until there's an engine fix? I'm not sure how embedded material definitions are and how much they can be modified, or if it would currently require a custom build (something I've never been able to do successfully :/ )

**Comment by Calinou** [comment]
This will be addressed by https://github.com/godotengine/godot/pull/50294 once it's fixed.

> Can materials be extended in their current form so I can get them to work like this until there's an engine fix?

No, you need to define a ShaderMaterial. You can convert a SpatialMaterial to a ShaderMaterial by using the toolbox icon in the inspector.

**Comment by elvisish** [comment]
Is this fixed for 3.4 RC?

**Comment by Calinou** [comment]
> Is this fixed for 3.4 RC?

No, see my above comment. This will only be fixed in 3.5 or later.

**Comment by elvisish** [comment]
Did this make it to 3.5?

**Comment by Calinou** [comment]
> Did this make it to 3.5?

No, it's too late for this change to make it to 3.5. You can create your own custom editor and export template binaries if you need this change in a project.

**Comment by elvisish** [comment]
Was this added to 3.5? I thought it was and just remembered it'll be its 1-year anniversary soon 🥹

**Comment by elvisish** [comment]
Could this planned to be backported to 3.6? It'll be it's 1-year anniversary soon! 🥹

**Comment by Calinou** [comment]
> Could this planned to be added to 3.6? It'll be it's 1-year anniversary soon! 🥹

https://github.com/godotengine/godot/pull/50294 needs to be reviewed and merged first. There are some concerns about orthogonal cameras which should be addressed in some way.

**Comment by elvisish** [comment]
> > Could this planned to be added to 3.6? It'll be it's 1-year anniversary soon! 🥹
> 
> #50294 needs to be reviewed and merged first. There are some concerns about orthogonal cameras which should be addressed in some way.

The merge seems blocked for the last 3 months?
